app/js/FSPersistorManager.js

```diff
@@ -1,206 +1,169 @@
-    })
-    const targetStream = fs.createWriteStream(`${location}/${filteredToName}`)
-    targetStream.on('error', function(err) {
-      logger.err(
-        { err, location, key: filteredToName },
-        'Error writing to file'
-      )
-      return callback(err)
-    })
-    targetStream.on('finish', () => callback(null))
-    return sourceStream.pipe(targetStream)
-  },
-
-  deleteFile(location, name, callback) {
-    const filteredName = filterName(name)
-    logger.log({ location, filteredName }, 'delete file')
-    return fs.unlink(`${location}/${filteredName}`, function(err) {
-      if (err != null) {
-        logger.err({ err, location, filteredName }, 'Error on delete.')
-        return callback(err)
-      } else {
-        return callback()
-      }
-    })
-  },
-
-  deleteDirectory(location, name, callback) {
-    if (callback == null) {
-      callback = function(err) {}
+    throw new ReadError('failed to stat file').withCause(err)
+  }
+}
+
+async function copyFile(location, fromName, toName) {
+  const filteredFromName = filterName(fromName)
+  const filteredToName = filterName(toName)
+  logger.log({ location, filteredFromName, filteredToName }, 'copying file')
+
+  const sourceStream = fs.createReadStream(`${location}/${filteredFromName}`)
+  const targetStream = fs.createWriteStream(`${location}/${filteredToName}`)
+  await pipeline(sourceStream, targetStream)
+}
+
+async function deleteFile(location, name) {
+  const filteredName = filterName(name)
+  logger.log({ location, filteredName }, 'delete file')
+  await fsUnlink(`${location}/${filteredName}`)
+}
+
+async function deleteDirectory(location, name) {
+  const filteredName = filterName(name.replace(/\/$/, ''))
+
+  logger.log({ location, filteredName }, 'deleting directory')
+
+  await rmrf(`${location}/${filteredName}`)
+}
+
+async function checkIfFileExists(location, name) {
+  const filteredName = filterName(name)
+  try {
+    const stat = await fsStat(`${location}/${filteredName}`)
+    return !!stat
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      return false
     }
-    const filteredName = filterName(name.replace(/\/$/, ''))
-    return rimraf(`${location}/${filteredName}`, function(err) {
-      if (err != null) {
-        logger.err({ err, location, filteredName }, 'Error on rimraf rmdir.')
-        return callback(err)
-      } else {
-        return callback()
-      }
-    })
-  },
-
-  checkIfFileExists(location, name, callback) {
-    if (callback == null) {
-      callback = function(err, exists) {}
+    throw new ReadError('failed to stat file').withCause(err)
+  }
+}
+
+// note, does not recurse into subdirectories
+async function directorySize(location, name) {
+  const filteredName = filterName(name.replace(/\/$/, ''))
+  let size = 0
+
+  try {
+    const files = await fsReaddir(`${location}/${filteredName}`)
+    for (const file of files) {
+      const stat = await fsStat(`${location}/${filteredName}/${file}`)
```

Comment:
It doesn't look like we're using this anywhere, but I don't know for sure as my grep assumes that we're building the string using interpolation, which might not be the case everywhere.

It's potentially useful and because it's not so easy to be sure I would prefer to leave it in.

Reply:
I should have been more specific here. I did not intend to drop it but check what type of projects are accessed here. Summing the size for immutable projects (e.g. template projects?) should be ok. Accessing active user projects however would be dangerous as the user could delete a file while we iterate through the files.

Reply:
On the other hand: I've tried adding an acceptance test for it.

it doesn't actually work with the FS Persistor, because it's looking for a directory and we don't have one. It works with S3 though. I still think it's better to keep it and fix the FS Persistor. I think it could be useful in the future.

Reply:
>  Accessing active user projects however would be dangerous as the user could delete a file while we iterate through the files.

Yes, that's true. If we handle not-found errors when `stat`ing the files this works around the problem. If the user does modify the project while we're doing this, the worst that will happen is that we get back an incorrect size. I don't think that's too much of a big deal, provided the not-found is correctly handled.